[Hub Generated] Review request for Microsoft.AlertsManagement to add version preview/2021-01-01-preview

dishonono opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/dishonono/azure-rest-api-specs/dev-alertsmanagement-Microsoft.AlertsManagement-2021-01-01-preview...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

dishonono committed:
Adds base for updating Microsoft.AlertsManagement from version preview/2020-08-04-preview to version 2021-01-01-preview

dishonono committed:
Updates readme

dishonono committed:
Updates API version in new specs and examples

dishonono committed:
first commit

dishonono committed:
fix: "Client Api Version."

dishonono committed:
try add  x-ms-long-running-operation-options

dishonono committed:
some renames

dishonono committed:
try to remove long operation

dishonono commented:
this is the ARM manifest PR:
https://armmanifest.visualstudio.com/One/_git/Manifest/pullrequest/2836

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,273 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
>"202" [](start = 10, length = 5)

You are missing a 200 response please refer to [this](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#call-action-post-asynchronously) doc, additionally you must add: "x-ms-long-running-operation": true like mentioned [here](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation)

dishonono replied:
I have added the x-ms-long-running-operation
however, why do I need 200? the post will not return 200, it will return in location the url for the get status (the get api in this swagger), and that one will return 200 . isn't that valid?

jorgecotillo replied:
Got it, your location is the one reporting the status, then no need for 200, what you have is ok.

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,273 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrateFromSmartDetection/{migrationId}": {
```
>"/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrateFromSmartDetection/{migrationId}" [](start = 4, length = 110)

You don't expose a list right?

dishonono replied:
not a list, why?

jorgecotillo replied:
Just making sure that a customer cannot return a list of migrations.

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/examples/migrateFromSmartDetectionStartMigration.json:
```diff
@@ -0,0 +1,24 @@
+  "parameters": {
+    "subscriptionId": "70afbb35-0463-4758-badf-d5aab5bce520",
+    "api-version": "2021-01-01-preview",
+    "parameters": {
```
>"parameters" [](start = 4, length = 12)

I believe this should be "properties" not parameters.

dishonono replied:
looking at other example files in the repo, there is paramters, and then another parameters, and only then properties as the 3rd level.
also the doc seem find with what i've done.
https://review.docs.microsoft.com/en-us/rest/api/documentation-preview/migratefromsmartdetection/startmigration?view=azure-rest-preview&branch=result_openapiHub_production_e62f880f06b4
are you sure?

dishonono replied:
Are you sure? looking in other examples in the repo, there are 2 levels of parameters and then properties comes as 3rd level (which I dont need).
It also looks fine in the doc: https://review.docs.microsoft.com/en-us/rest/api/documentation-preview/migratefromsmartdetection/startmigration?view=azure-rest-preview&branch=result_openapiHub_production_e62f880f06b4

jorgecotillo replied:
I see, it works because you have "parameters" as name of your POST body:

```
"name": "parameters",
```

Which is a little bit odd, because usually the property holding the request body has a name related to the operation, something like "migration" in your case or something like that.

---
In reply to: [570234800](https://github.com/Azure/azure-rest-api-specs/pull/12748#discussion_r570234800) [](ancestors = 570234800)

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,273 @@
+        "tags": [
+          "MigrateFromSmartDetection"
+        ],
+        "description": "Retrieve the status of a migration process.",
```
>Retrieve the status of a migration process. [](start = 24, length = 43)

How does a user know the migrationId?

jorgecotillo replied:
For long-running operations follow this RPC: https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#call-action-post-asynchronously

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,273 @@
+      "required": [
+        "scope"
+      ],
+      "properties": {
```
>"properties" [](start = 6, length = 12)

Are all your properties optional? otherwise add "required" like you have done below.

dishonono replied:
yes, only scope is mandatory in the request

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,273 @@
+        "scope"
+      ],
+      "properties": {
+        "id": {
```
>"id" [](start = 8, length = 4)

You should use a different value than "id", id should be used for fully qualified Azure resource ids. If this is a proxy resource, follow [this](https://github.com/Azure/azure-rest-api-specs/blob/b6612fae3975f2396339b239914d5394a320c999/specification/common-types/resource-management/v1/types.json#L82) definition, for tracked resources follow [this](https://github.com/Azure/azure-rest-api-specs/blob/b6612fae3975f2396339b239914d5394a320c999/specification/common-types/resource-management/v1/types.json#L48) instead.

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/examples/migrateFromSmartDetectionStartMigration.json:
```diff
@@ -0,0 +1,24 @@
+    "202": {
+      "headers": {},
+      "body": {
+        "id": "ebf4f1bf-2f56-44b2-99b0-870867ae549b",
```
>"id" [](start = 8, length = 4)

Your example should return an "id", "name" and "type" for any Azure resource

dishonono replied:
ok. so I must return an Azure Resource?
the Id of this API is to allow users to trigger a migration process (takes couple of minutes) and then get there migration status. it's a proxy resource. and once they migration is completed we don't intend to treat this as resources, that you can delete etc...
I can return type, id and put the id that is currently returned in name. although migrationId would be clearer than name.

jorgecotillo replied:
at minimum "id" and "type" should be included, "id" should relate to your Azure resource Id and not your custom migrationId, you can expose migrationId inside your "properties"

dishonono committed:
fixes from code review

dishonono commented:
/azp run unifiedPipeline

dishonono committed:
fix typo

dishonono commented:
/azp run unifiedPipeline

dishonono committed:
code review fixes

dishonono committed:
try fix lintdiff

dishonono commented:
Hi @jorgecotillo ,
I've fixed the comments you had.
one issue is that because I added x-ms-long-running-operation to the post, I now see this:
![image](https://user-images.githubusercontent.com/30979046/107142388-449a1b00-6937-11eb-994d-80d98d3c33f5.png)
Although, as I said, the final status 200 will be delivered by the GET operation that the user is referred to through the location header.

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,304 @@
+            "$ref": "./examples/migrateFromSmartDetectionStartMigration.json"
+          }
+        },
+        "x-ms-long-running-operation": true,
```
>"x-ms-long-running-operation": true, [](start = 8, length = 36)

According to our RPC, you still need to return a terminal state - that's why you are getting the linting error, see [here](https://github.com/Azure/azure-rest-api-specs/blob/44ec11c74634f1afa5bbe0d3f2678b7f819ebf8e/specification/apimanagement/control-plane/Microsoft.ApiManagement/preview/2017-03-01/apimtenant.json#L340) to find an example. 

The way it should work is as follows:

1. First POST call returns 202
2. Location header contains a URL to monitor the operation progress
3. If another POST call is invoked to the same resource id (URL path) using the same exact body, your service should treat it as a no-op, returns a response code = 200 and the response content should be your resource object - this is the reason why you need to return a terminal state

Hope it makes sense.

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,304 @@
+      "type": "object",
+      "properties": {
+        "id": {
+          "description": "The migration process ID.",
```
>"description": "The migration process ID.", [](start = 10, length = 43)

This is no longer your migration process id right? It is now: "Migration resource identifier"

jorgecotillo commented on specification/alertsmanagement/resource-manager/Microsoft.AlertsManagement/preview/2021-01-01-preview/MigrateFromSmartDetections.json:
```diff
@@ -0,0 +1,304 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrateFromSmartDetection/{migrationId}": {
```
>"/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrateFromSmartDetection/{migrationId}" [](start = 4, length = 110)

Just had a second thought here, do you still need this endpoint? your POST action seems similar to this one, POST action should expose an action name: e.g. restart, there's nothing else after the action name.

You can model this like this:

"/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrations" -> to return a list

"/subscriptions/{subscriptionId}/providers/Microsoft.AlertsManagement/migrations/{migrationId}" -> to return a single item

jorgecotillo commented:
@dishonono take a look at my GET and POST comment from 2/10

dishonono committed:
remove the GET from the swagger , Add 200 to the POST

dishonono committed:
fix spelling

dishonono commented:
/azp run unifiedPipeline

dishonono committed:
remove location header from the 200 response

dishonono commented:
/azp run unifiedPipeline

akning-ms commented on specification/alertsmanagement/resource-manager/readme.md:
```diff
@@ -44,6 +44,16 @@ openapi-type: arm
 tag: package-2019-06-preview
```
Will you use new api version for your SDK release? if yes, we should change tag to your new tag for 2021-01 version

dishonono replied:
Hi,
no . We don't want to use the new API version for the SDK release
